File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/interval_intersection.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval() : start(-1), end(-1) {}

  Interval(int start, int end) : start(start), end(end) {}
};

// O(n + m) / O(1)
vector<Interval> intersection(vector<Interval> &a, vector<Interval> &b) {
  vector<Interval> res;
  int i = 0, j = 0;

  auto has_overlap = [](const auto &e1, const auto &e2) {
    return (e1.start >= e2.start && e1.start <= e2.end) ||
           (e2.start >= e1.start && e2.start <= e1.end);
  };

  while (i < a.size() && j < b.size()) {
    if (has_overlap(a[i], b[j])) {
      int start = max(a[i].start, b[j].start);
      int end = min(a[i].end, b[j].end);
      res.push_back(Interval(start, end));
    }

    if (a[i].end < b[j].end) {
      i++;
    } else {
      j++;
    }
  }

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, m, start, end;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a(n);
    for (int i = 0; i < n; i++) {
      cin >> start >> end;
      a[i] = Interval(start, end);
    }

    cin >> m;
    vector<Interval> b(m);
    for (int i = 0; i < m; i++) {
      cin >> start >> end;
      b[i] = Interval(start, end);
    }

    auto res = intersection(a, b);
    for (auto r : res) {
      cout << r.start << " " << r.end << "\n";
    }
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/IntervalIntersection.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class IntervalIntersection {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          a.add(new Interval(start, end));
        }

        int m = in.nextInt();
        List<Interval> b = new ArrayList<>(m);
        for (int i = 0; i < m; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          b.add(new Interval(start, end));
        }

        List<Interval> res = intersection(a, b);
        for (var r : res) {
          System.out.printf("%d %d\n", r.start, r.end);
        }
      }
    }
  }

  // O(n + m) / O(1)
  public static List<Interval> intersection(List<Interval> a,
                                            List<Interval> b) {
    List<Interval> res = new ArrayList<>();
    int i = 0, j = 0;

    BiPredicate<Interval, Interval> hasOverlap = (x, y) -> {
      return (x.start >= y.start && x.start <= y.end) ||
          (y.start >= x.start && y.start <= x.end);
    };

    while (i < a.size() && j < b.size()) {
      if (hasOverlap.test(a.get(i), b.get(j))) {
        int start = Math.max(a.get(i).start, b.get(j).start);
        int end = Math.min(a.get(i).end, b.get(j).end);

        res.add(new Interval(start, end));
      }

      if (a.get(i).end < b.get(j).end) {
        i++;
      } else {
        j++;
      }
    }

    return res;
  }
}